source/common/common/perf_annotation.h

```diff
@@ -0,0 +1,162 @@
+ */
+#define PERF_RECORD(perf, category, description)                                                   \
+  do {                                                                                             \
+    perf.record(category, description);                                                            \
+  } while (false)
+
+/**
+ * Dumps recorded performance data to stdout. Expands to nothing if not enabled.
+ */
+#define PERF_DUMP() Envoy::PerfAnnotationContext::dump()
+
+/**
+ * Returns the aggregated performance data as a formatted multi-line string, showing a
+ * formatted table of values. Returns "" if perf-annotation is disabled.
+ */
+#define PERF_TO_STRING() Envoy::PerfAnnotationContext::toString()
+
+/**
+ * Clears all performance data.
+ */
+#define PERF_CLEAR() Envoy::PerfAnnotationContext::clear()
+
+/**
+ * Controls whether performacne collection and reporting is thread safe. For now,
+ * leaving this enabled for predictability across multiiple applications, on the assumption
+ * that an uncontended mutex lock has vanishingly small cost. In the future we may try
+ * to make this system thread-unsafe if mutext contention disturbs the metrics.
+ */
+#define PERF_THREAD_SAFE true
+
+namespace Envoy {
+
+/**
+ * Defines a context for collecting performance data. Note that this class is
+ * fully declared and defined even if ENVOY_PERF_AUTOMATION is off. We depend on
+ * the macros to disable performance collection for production.
+ */
+class PerfAnnotationContext {
+public:
+  /**
+   * Records time consumed by a category and description, which are just
+   * joined together in the library with " / ".
+   */
+  void record(std::chrono::nanoseconds duration, absl::string_view category,
+              absl::string_view description);
+
+  /**
+   * Renders the aggregated statistics as a string.
+   * @return std::string the performance data as a formatted string.
+   */
+  static std::string toString();
+
+  /**
+   * Dumps aggregated statistics (if any) to stdout.
+   */
+  static void dump();
+
+  /**
+   * Thread-safe lazy-initialization of a PerfAnnotationContext on first use.
+   * @return PerfAnnotationContext* the context.
+   */
+  static PerfAnnotationContext* getOrCreate();
+
+  /**
+   * Clears out all aggregated statistics.
+   */
+  static void clear();
+
+private:
+  /**
+   * PerfAnnotationContext construction should be done via getOrCreate().
+   */
+  PerfAnnotationContext();
+
+  typedef std::pair<std::chrono::nanoseconds, uint64_t> DurationCount;
+  typedef std::map<std::string, DurationCount> DurationCountMap;
+
+  DurationCountMap duration_count_map_; // Maps "$category / $description" to the duration.
+#if PERF_THREAD_SAFE
+  std::mutex mutex_;
+#endif
+};
+
+/**
+ * Represents an operation for reporting timing to the perf system. Usage:
+ *
+ * f() {
+ *   PerfOperation perf_op;
+ *   computeIntensiveWork();
+ *   perf_op.record("category", "description");
+ * }
+ */
+class PerfOperation {
```

Comment:
There don't appear to be any tests for this class below.  Can we add a few?  I think just using a second timer to ensure that the recorded time will be greater than some value will suffice.